File: matrix-c-plus/matrix_multiplication.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <omp.h>

using namespace std;

void transposeMatrix(const vector<vector<int>> &matrix, vector<vector<int>> &transposed, int size) {
    #pragma omp parallel for collapse(2)
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            transposed[j][i] = matrix[i][j];
        }
    }
}

void multiplyMatrices(const vector<vector<int>> &a, const vector<vector<int>> &b, vector<vector<int>> &result, int size) {
    vector<vector<int>> bTransposed(size, vector<int>(size, 0));
    transposeMatrix(b, bTransposed, size);

    #pragma omp parallel for collapse(2)
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            result[i][j] = 0;
            for (int k = 0; k < size; ++k) {
                result[i][j] += a[i][k] * bTransposed[j][k];
            }
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cout << "Please provide the matrix size as an argument." << endl;
        return 1;
    }

    int size = stoi(argv[1]);

    vector<vector<int>> a(size, vector<int>(size, 1));
    vector<vector<int>> b(size, vector<int>(size, 1));
    vector<vector<int>> result(size, vector<int>(size, 0));

    auto start = chrono::high_resolution_clock::now();
    multiplyMatrices(a, b, result, size);
    auto end = chrono::high_resolution_clock::now();

    chrono::duration<double> elapsed = end - start;
    cout << "Matrix multiplication took: " << elapsed.count() << " seconds." << endl;

    return 0;
}

File: matrix-rust/matrix_multiplication.rs
use rayon::prelude::*;
use std::env;
use std::time::Instant;

fn transpose_matrix(matrix: &Vec<Vec<i32>>, size: usize) -> Vec<Vec<i32>> {
    let mut transposed = vec![vec![0; size]; size];
    for i in 0..size {
        for j in 0..size {
            transposed[j][i] = matrix[i][j];
        }
    }
    transposed
}

fn multiply_matrices(a: &Vec<Vec<i32>>, b: &Vec<Vec<i32>>, result: &mut Vec<Vec<i32>>, size: usize) {
    let b_transposed = transpose_matrix(b, size);
    result.par_iter_mut().enumerate().for_each(|(i, row)| {
        for j in 0..size {
            row[j] = 0;
            for k in 0..size {
                row[j] += a[i][k] * b_transposed[j][k];
            }
        }
    });
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Please provide the matrix size as an argument.");
        return;
    }

    let size: usize = args[1].parse().expect("Invalid matrix size");

    let a = vec![vec![1; size]; size];
    let b = vec![vec![1; size]; size];
    let mut result = vec![vec![0; size]; size];

    let start = Instant::now();
    multiply_matrices(&a, &b, &mut result, size);
    let duration = start.elapsed();

    println!("Matrix multiplication took: {:?}", duration);
}
